reps/2022-12-08-ray-for-federated-learning-and-privacy-preserving-computing.md

```diff
@@ -0,0 +1,167 @@
+```
+
+### Understanding the mechanism
+In this section, we're introducing the deep dive of the above federated learning example, to help understand the mechanism of this connector layer.
+
+Firstly,  ALICE and BOB companies need to create their Ray clusters, and expose one communication port to each other.  
+And then, both ALICE and BOB need to run the example code in their Ray clusters(drive the DAG in multi-controller mode). Note that, the peer port and parties info should be passed via `fed.init()`.
+![image](https://user-images.githubusercontent.com/19473861/206469580-468ea258-407e-40cf-b059-06a63a25c168.png)
+
+Every Ray cluster will create the Ray tasks specified to its party, meanwhile, it will send the output data  to peer if the downstream Ray task are not specified to its own party. For example, in ALICE cluster, `mean.party('alice').remote(w_a, w_b)` will create a Ray task in ALICE cluster, but it wouldn't be created in BOB cluster. Due to `w_b` is an output data from BOB cluster, therefore, in ALICE cluster, it will be inserted a `recv_op` barrier to `mean` task, and in BOB cluster, it will be inserted a `send_op` barrier as a downstream of `w_b = my_bob.get_weights.remote()` task.
+
+### Benefits 
+Compared with running the DAG in one Ray cluster, we have the following significant benefits:
+- It's very close to Ray native programming pattern. But the DAG could be ran across Ray clusters.
+- Very restricted and clear data perimeters. we only transmit the data which is needed by others. And all of the data couldn't be fetched in any way if we don't allow.
+- If we run the DAG in one Ray cluster, data transmition is in PULL-BASED mode. For example, if BOB invokes `ray.get(f.options("PARTY_ALICE").remote())`, the return object is pulled by BOB, as a result, ALICE don't have the knowledge for that. In this proposal, it's in PUSH-BASED mode. ALICE has the knowledge for that there is a data object will be sent to BOB. This is a significant advantage of multi-controller mode.
+- All the disadvanteges are addressed in this proposal: code distribution, data privacy, task-attacks, illegal invasion, deserialization vulnerability，and etc.
+- Brings all Ray advantages to the local Ray cluster, like highly performance RPC, fault tolerance, task scheduling/resource management, and other Ray ecosystem libraries（ Ray datasets, Ray train and Ray serve） for local training.
+
+
+
+## Compatibility, Deprecation, and Migration Plan
+None.
+
+## Test Plan and Acceptance Criteria
+- Test should includes unit tests and integration tests(release tests).
+- Tests should be enabled in CI.
+- Benchmark on across parties data transmition for typical federated learning case.
+- Document page and necessary code comments.
+
+## (Optional) Follow-on Work
+- UX improvements for across parties exceptions passing.
+- Provide single-controller mode for quickly building applications and debug inner one party.
+- Performance optimization for other workloads.
```

Comment:
Maybe 2 more items should be added in this section:
- A unified gateway for ETH communication.
- Support TEE(Trust Execution Environment) device.

Reply:
Yea, nice suggestions.